### PR TITLE
SQL: more Fluent updates

### PR DESCRIPTION
**Pull request by milismsft**
Add Fluent implementation and tests for SQL Encryption Protectors, Sync Groups and Sync Members

### CODE REVIEW

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        /// <param name="sqlServerManager">Reference to the SQL server manager that accesses firewall rule operations.</param>
+        ///GENMHASH:4DD6445361F09E0B185DE0E8D24170D3:81D0EDB620000F6B86DC1403C1548E6A
+        internal SqlEncryptionProtectorImpl(EncryptionProtectorInner innerObject, ISqlManager sqlServerManager)
+            : base(innerObject, null)
+        {
+            this.sqlServerManager = sqlServerManager;
+            if (innerObject != null && innerObject.Id != null)
+            {
+                if (innerObject.Id != null)
+                {
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: you already check for this not being null 2 lines above, right?

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        }
+
+        ///GENMHASH:39E79EEEBE800C0263D7785D2CDD0C8F:AA25005E7A9BAF367EA281FE0DB49192
+        public string Uri()
+        {
+            return this.Inner.Uri;
+        }
+
+        ///GENMHASH:5AD91481A0966B059A478CD4E9DD9466:2EA6F22A1E9075415416C51D51B2ED7C
+        protected async Task<Models.EncryptionProtectorInner> GetInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: shouldn't this be an overriden method?

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            return await this.sqlServerManager.Inner.EncryptionProtectors
+                .GetAsync(this.resourceGroupName, this.sqlServerName, cancellationToken);
+        }
+
+        ///GENMHASH:7A0398C4BB6EBF42CC817EE638D40E9C:2DC6B3BEB4C8A428A0339820143BFEB3
+        public string ParentId()
+        {
+            var resourceId = ResourceId.FromString(this.Id());
+            return resourceId.Parent.Id;
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: would be safer to do `return resourceId?.Parent?.Id;`

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+        ///GENMHASH:6A2970A94B2DD4A859B00B9B9D9691AD:6475F0E6B085A35B081FA09FFCBDDBF8
+        public Region Region()
+        {
+            return Microsoft.Azure.Management.ResourceManager.Fluent.Core.Region.Create(this.Inner.Location);
+        }
+
+        ///GENMHASH:507A92D4DCD93CE9595A78198DEBDFCF:4C4C4336C86119672D7AD1E0D4BD29CB
+        public async Task<Microsoft.Azure.Management.Sql.Fluent.ISqlEncryptionProtector> UpdateResourceAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: shouldn't this be overriden?

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            return await CreateResourceAsync(cancellationToken);
+        }
+
+        ///GENMHASH:0202A00A1DCF248D2647DBDBEF2CA865:06F8F4D881AC3D94F49CECB6153B5A83
+        public async Task<Microsoft.Azure.Management.Sql.Fluent.ISqlEncryptionProtector> CreateResourceAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: shouldn't this be overriden?

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            return this;
+        }
+
+        public ISqlEncryptionProtector Apply()
+        {
+            return Extensions.Synchronize(() => this.ApplyAsync());
+        }
+
+        public async Task<ISqlEncryptionProtector> ApplyAsync(CancellationToken cancellationToken = default(CancellationToken), bool multiThreaded = true)
+        {
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: overriden?

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            return this;
+        }
+
+        public ISqlEncryptionProtector Refresh()
+        {
+            return Extensions.Synchronize(() => this.RefreshAsync());
+        }
+
+        public async Task<ISqlEncryptionProtector> RefreshAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: overriden?

File: src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            return Extensions.Synchronize(() => this.GetByIdAsync(id));
+        }
+
+        ///GENMHASH:90DDF31DBCC18A1CCF68558A7E8449CD:322C7F5994350DD787C93CBDCD3773CB
+        public async Task<IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.ISqlEncryptionProtector>> ListBySqlServerAsync(string resourceGroupName, string sqlServerName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            List<Microsoft.Azure.Management.Sql.Fluent.ISqlEncryptionProtector> encryptionProtectors = new List<ISqlEncryptionProtector>();
+            var encryptionProtectorInners = await this.sqlServerManager.Inner.EncryptionProtectors
+                .ListByServerAsync(resourceGroupName, sqlServerName);
+            if (encryptionProtectorInners != null)
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs]: missed cancellationToken

File: src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            return encryptionProtectors.AsReadOnly();
+        }
+
+        ///GENMHASH:2EA8D9073369A6E1267C1FB74F86952A:194FD06C0A433DA71B2E0D06F4181B3E
+        public async Task<IReadOnlyList<Microsoft.Azure.Management.Sql.Fluent.ISqlEncryptionProtector>> ListBySqlServerAsync(ISqlServer sqlServer, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            List<Microsoft.Azure.Management.Sql.Fluent.ISqlEncryptionProtector> encryptionProtectors = new List<ISqlEncryptionProtector>();
+            var encryptionProtectorInners = await this.sqlServerManager.Inner.EncryptionProtectors
+                .ListByServerAsync(sqlServer.ResourceGroupName, sqlServer.Name);
+            if (encryptionProtectorInners != null)
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs]: missed cancellationToken

File: src/ResourceManagement/Sql/SqlSyncGroupImpl.cs [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+        /// <param name="sqlServerManager">Reference to the SQL server manager that accesses DNS alias operations.</param>
+        ///GENMHASH:2593A87C1076A6D668BDFE33052F341C:E7C65DE6BF45CF0BA29C2A32A669F3CF
+        internal SqlSyncGroupImpl(string name, SyncGroupInner innerObject, ISqlManager sqlServerManager)
+            : base(innerObject, null)
+        {
+            this.name = name;
+            this.sqlServerManager = sqlServerManager;
+            if (innerObject != null && innerObject.Id != null)
+            {
+                if (innerObject.Id != null)
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlSyncGroupImpl.cs]: looks like redundant check (line 91 already has it)

File: src/ResourceManagement/Sql/SqlSyncMemberImpl.cs [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+        ///GENMHASH:DE04C9A4A480ADDBCD794C715F4CCE5C:657722CF28579ADFAAA5E7E0CF6B5310
+        internal SqlSyncMemberImpl(string name, SyncMemberInner innerObject, ISqlManager sqlServerManager)
+            : base(innerObject, null)
+        {
+            this.name = name;
+            this.sqlServerManager = sqlServerManager;
+            if (innerObject != null && innerObject.Id != null)
+            {
+                if (innerObject.Id != null)
+                {
```
Comment (hovsepm) [inline src/ResourceManagement/Sql/SqlSyncMemberImpl.cs]: redundant copy paste :)

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs (hunk 1 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: Fixed

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs (hunk 2 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: not in this case; the Impl inherits from ChildResource<> which does not define it.

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs (hunk 4 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: Same comment as before, this class inherits from ChildSource<> which does not define this method. This method comes as part of the Java2CSharp conversion.

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs (hunk 5 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: Same comment as before, this class inherits from ChildSource<> which does not define this method. This method comes as part of the Java2CSharp conversion.

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs (hunk 6 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: nope; same comment as above.

File: src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs (hunk 7 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorImpl.cs]: nope, same comment as above.

File: src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs (hunk 9 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs]: fixed

File: src/ResourceManagement/Sql/SqlSyncGroupImpl.cs (hunk 10 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlSyncGroupImpl.cs]: fixed

File: src/ResourceManagement/Sql/SqlSyncMemberImpl.cs (hunk 11 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlSyncMemberImpl.cs]: fixed

File: src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs (hunk 8 above)
Comment (milismsft) [reply src/ResourceManagement/Sql/SqlEncryptionProtectorOperationsImpl.cs]: forgot about this one; fixed.

**Pull request merged**